File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch08/C/P08-Roothaan.cpp
// Solves the Roothaan equation F C = S C E for the H2O molecule
#include <stdio.h>
#include "eigsys.h"
#include "matutil.h"

int main()
{
   double **C, **F, **S, **S0;
   double *E, err, t;
   int i, j, k, n;
   FILE *f;

   n = 13;

   C = Matrix(1,n,1,n);                             // molecular coefficients
   F = Matrix(1,n,1,n);                                        // Fock matrix
   S = Matrix(1,n,1,n);                                     // overlap matrix
   S0 = Matrix(1,n,1,n);                     // backup copy of overlap matrix
   E = Vector(1,n);                                          // energy levels

   f = fopen("H2O.dat","r");                               // read in F and S
   if (f == NULL) { printf(" H2O.dat missing !\n"); exit(1); }
   while(fgetc(f) != '\n');                                    // skip header
   for (i=1; i<=n; i++)                        // read in lower triangle of F
      for (j=1; j<=i; j++) {                       // F not altered by EigSym
         fscanf(f,"%lf",&F[i][j]);
         F[j][i] = F[i][j];                                     // symmetrize
      }
   while(fgetc(f) != '\n');                               // finish last line

   while(fgetc(f) != '\n');                                    // skip header
   for (i=1; i<=n; i++)                        // read in lower triangle of S
      for (j=1; j<=i; j++) {                         // S destroyed by EigSym
         fscanf(f,"%lf",&S[i][j]);
         S[j][i] = S[i][j];                                     // symmetrize
         S0[i][j] = S0[j][i] = S[i][j];                           // backup S
      }
   fclose(f);

   printf("F: Fock matrix (sample)\n");
   MatPrint(F,5,5);
   printf("S: overlap matrix (sample)\n");
   MatPrint(S,5,5);

   EigSym(F,S,C,E,n);                              // solve Roothaan equation
   EigSort(C,E,n,1);                     // sort eigenvalues and eigenvectors

   printf("\nEigenenergies:\n");
   VecPrint(E,n);

   printf("\nHOMO-LUMO gap E[6]-E[5] = %4.3f Hartrees\n",E[6]-E[5]);

   err = 0e0;
   for (i=1; i<=n; i++)
      for (j=1; j<=n; j++) {
         t = 0e0;
         for (k=1; k<=n; k++) t += (F[i][k] - E[j]*S0[i][k]) * C[k][j];
         if (err < fabs(t)) err = fabs(t);       // err = max|a x - lambda x|
      }
   printf("\nMaximum error = %9.2e\n",err);
}

File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch08/Python/P08-Roothaan.py
# Solves the Roothaan equation F C = S C E for the H2O molecule
from eigsys import *
from matutil import *

n = 13

C = [[0]*(n+1) for i in range(n+1)]                  # molecular coefficients
F = [[0]*(n+1) for i in range(n+1)]                             # Fock matrix
S = [[0]*(n+1) for i in range(n+1)]                          # overlap matrix
S0 = [[0]*(n+1) for i in range(n+1)]          # backup copy of overlap matrix
E = [0]*(n+1)                                                 # energy levels

f = open('H2O.dat','r')                                     # read in F and S
line = f.readline()                                             # skip header
for i in range(1,n+1):                          # read in lower triangle of F
   line = f.readline()                              # F not altered by EigSym
   for j in range(1,i+1):
      F[i][j] = float(line.split()[j-1])
      F[j][i] = F[i][j]                                          # symmetrize

line = f.readline()                                             # skip header
for i in range(1,n+1):
   line = f.readline()
   for j in range(1,i+1):                       # read in lower triangle of S
      S[i][j] = float(line.split()[j-1])              # S destroyed by EigSym
      S0[i][j] = S0[j][i] = S[i][j]                                # backup S
f.close()

print("F: Fock matrix (sample)")
MatPrint(F,5,5)
print("S: overlap matrix (sample)")
MatPrint(S,5,5)

EigSym(F,S,C,E,n)                                   # solve Roothaan equation
EigSort(C,E,n,1)                          # sort eigenvalues and eigenvectors

print("\nEigenenergies:")
VecPrint(E,n)

print("\nHOMO-LUMO gap E[6]-E[5] = {0:4.3f} Hartrees".format(E[6]-E[5]))

err = 0e0
for i in range(1,n+1):
   for j in range(1,n+1):
      t = 0e0
      for k in range(1,n+1): t += (F[i][k] - E[j]*S0[i][k]) * C[k][j]
      if (err < fabs(t)): err = fabs(t)           # err = max|a x - lambda x|

print("\nMaximum error = {0:8.2e}".format(err))
